File: tide_offline.cpp
// tide_offline.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <nlohmann/json.hpp>
#include <getopt.h>

using namespace std;
using json = nlohmann::json;

const double DEFAULT_CONSTITUENTS[][3] = {
    {1.2, 12.42, 0.0},
    {0.4, 12.00, 0.0},
    {0.2, 12.66, 0.0},
    {0.1, 23.93, 0.0},
    {0.08, 25.82, 0.0}
};
const double MEAN_LEVEL = 0.5;

class TideModel {
private:
    vector<array<double,3>> constituents;
    double meanLevel;
public:
    TideModel(const vector<array<double,3>>& cons, double mean) : constituents(cons), meanLevel(mean) {}

    double level(double hours) const {
        double lvl = meanLevel;
        for (auto& c : constituents) {
            double amp = c[0], period = c[1], phase = c[2];
            double phaseRad = phase * M_PI / 180.0;
            double omega = 2 * M_PI / period;
            lvl += amp * sin(omega * hours + phaseRad);
        }
        return lvl;
    }

    vector<pair<double,double>> findExtrema(double start, double duration) const {
        double step = 0.05;
        vector<double> times, levels;
        for (double t = start; t <= start + duration; t += step) {
            times.push_back(t);
            levels.push_back(level(t));
        }
        vector<pair<double,double>> extrema;
        for (size_t i = 1; i < times.size()-1; i++) {
            if (levels[i] > levels[i-1] && levels[i] > levels[i+1])
                extrema.push_back({times[i], levels[i]});
            else if (levels[i] < levels[i-1] && levels[i] < levels[i+1])
                extrema.push_back({times[i], levels[i]});
        }
        return extrema;
    }
};

double hoursSinceEpoch(const tm& dt) {
    time_t epoch = 946684800; // 2000-01-01 00:00:00 UTC
    tm epoch_tm = *gmtime(&epoch);
    time_t t = mktime(const_cast<tm*>(&dt));
    return difftime(t, epoch) / 3600.0;
}

string formatTime(double hours) {
    int h = (int)hours % 24;
    int m = (int)((hours - h) * 60);
    char buf[6];
    snprintf(buf, sizeof(buf), "%02d:%02d", h, m);
    return string(buf);
}

string drawGraph(const vector<double>& levels, double start, double duration, int width=50, int height=20) {
    if (levels.empty()) return "No data";
    double minLvl = *min_element(levels.begin(), levels.end());
    double maxLvl = *max_element(levels.begin(), levels.end());
    double rangeLvl = maxLvl - minLvl;
    if (rangeLvl == 0) rangeLvl = 1;
    int cols = width;
    double step = duration / cols;
    vector<double> sampled;
    for (int i=0; i<cols; i++) {
        double t = start + i * step;
        int idx = (int)((t - start) / step);
        if (idx >= (int)levels.size()) idx = levels.size()-1;
        sampled.push_back(levels[idx]);
    }
    double minS = *min_element(sampled.begin(), sampled.end());
    double maxS = *max_element(sampled.begin(), sampled.end());
    double rangeS = maxS - minS;
    if (rangeS == 0) rangeS = 1;
    vector<vector<char>> grid(height, vector<char>(cols, ' '));
    for (int i=0; i<cols; i++) {
        int row = (int)((sampled[i] - minS) / rangeS * (height - 1));
        if (row < 0) row = 0;
        if (row >= height) row = height - 1;
        grid[row][i] = '*';
    }
    stringstream ss;
    ss << "Level (m)\n";
    for (int r = height-1; r >= 0; r--) {
        double lvl = minS + (maxS - minS) * r / (height - 1);
        ss << fixed << setprecision(2) << setw(6) << lvl << " |";
        for (int c=0; c<cols; c++) ss << grid[r][c];
        ss << "\n";
    }
    ss << "      +" << string(cols-1, '-') << "\n";
    ss << "       ";
    for (int h=0; h<=duration; h+=3) {
        int pos = (int)(h / duration * cols);
        string label = (h<10?"0":"") + to_string(h) + ":00";
        while ((int)ss.str().length() < pos) ss << " ";
        ss << label;
    }
    ss << "\n";
    return ss.str();
}

int main(int argc, char* argv[]) {
    static struct option long_options[] = {
        {"date", required_argument, 0, 'd'},
        {"hours", required_argument, 0, 'h'},
        {"graph", no_argument, 0, 'g'},
        {"list", no_argument, 0, 'l'},
        {"config", required_argument, 0, 'c'},
        {"save-config", required_argument, 0, 's'},
        {0,0,0,0}
    };
    int opt;
    string dateStr, configFile, saveConfigFile;
    double hours = 24.0;
    bool graph = false, list = false;
    while ((opt = getopt_long(argc, argv, "d:h:glc:s:", long_options, nullptr)) != -1) {
        switch (opt) {
            case 'd': dateStr = optarg; break;
            case 'h': hours = stod(optarg); break;
            case 'g': graph = true; break;
            case 'l': list = true; break;
            case 'c': configFile = optarg; break;
            case 's': saveConfigFile = optarg; break;
            default:
                cerr << "Usage: tide_offline --date YYYY-MM-DD --hours N --graph --list --config FILE --save-config FILE\n";
                return 1;
        }
    }

    vector<array<double,3>> constituents;
    for (auto& c : DEFAULT_CONSTITUENTS) constituents.push_back({c[0], c[1], c[2]});
    double meanLevel = MEAN_LEVEL;

    if (!configFile.empty()) {
        ifstream f(configFile);
        if (!f.is_open()) { cerr << "Cannot open config file\n"; return 1; }
        json j;
        f >> j;
        if (j.contains("constituents")) {
            constituents.clear();
            for (auto& item : j["constituents"]) {
                constituents.push_back({item[0], item[1], item[2]});
            }
        }
        if (j.contains("mean_level")) meanLevel = j["mean_level"];
    }

    if (!saveConfigFile.empty()) {
        json j;
        j["constituents"] = json::array();
        for (auto& c : constituents) {
            j["constituents"].push_back({c[0], c[1], c[2]});
        }
        j["mean_level"] = meanLevel;
        ofstream f(saveConfigFile);
        f << setw(2) << j << endl;
        cout << "Config saved to " << saveConfigFile << "\n";
        return 0;
    }

    time_t now = time(nullptr);
    tm dt = *gmtime(&now);
    if (!dateStr.empty()) {
        strptime(dateStr.c_str(), "%Y-%m-%d", &dt);
    }
    dt.tm_hour = 0; dt.tm_min = 0; dt.tm_sec = 0;

    double startHours = hoursSinceEpoch(dt);
    TideModel model(constituents, meanLevel);
    auto extrema = model.findExtrema(startHours, hours);

    if (graph) {
        vector<double> levels;
        double step = 0.1;
        for (double t = startHours; t <= startHours + hours; t += step)
            levels.push_back(model.level(t));
        cout << drawGraph(levels, startHours, hours);
    } else {
        cout << "\n🌊 Offline Tide Calendar\n";
        char dateBuf[20];
        strftime(dateBuf, sizeof(dateBuf), "%Y-%m-%d %H:%M", &dt);
        cout << "Date: " << dateBuf << " – " << hours << "h forecast\n\n";
        cout << "High/Low Tides:\n";
        sort(extrema.begin(), extrema.end());
        for (auto& e : extrema) {
            if (e.first >= startHours && e.first <= startHours + hours) {
                time_t t = mktime(&dt) + (time_t)((e.first - startHours) * 3600);
                tm evt = *gmtime(&t);
                char buf[20];
                strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M", &evt);
                string typ = (e.second > meanLevel) ? "High" : "Low";
                cout << buf << "  " << typ << "    " << fixed << setprecision(2) << e.second << " m\n";
            }
        }
    }
    return 0;
}

File: tide_offline.py
# tide_offline.py
import sys
import math
import json
import argparse
from datetime import datetime, timedelta, timezone

# Default harmonic constituents: [(amplitude_m, period_hours, phase_deg)]
DEFAULT_CONSTITUENTS = [
    (1.2, 12.42, 0.0),   # M2
    (0.4, 12.00, 0.0),   # S2
    (0.2, 12.66, 0.0),   # N2
    (0.1, 23.93, 0.0),   # K1
    (0.08, 25.82, 0.0),  # O1
]
MEAN_LEVEL = 0.5  # meters

class TideModel:
    def __init__(self, constituents=None, mean_level=MEAN_LEVEL):
        self.constituents = constituents or DEFAULT_CONSTITUENTS
        self.mean_level = mean_level

    def level(self, hours_since_epoch):
        level = self.mean_level
        for amp, period, phase in self.constituents:
            phase_rad = phase * math.pi / 180.0
            omega = 2 * math.pi / period
            level += amp * math.sin(omega * hours_since_epoch + phase_rad)
        return level

    def find_extrema(self, start_hours, duration, step=0.05):
        times = []
        levels = []
        t = start_hours
        while t <= start_hours + duration:
            times.append(t)
            levels.append(self.level(t))
            t += step
        highs = []
        lows = []
        for i in range(1, len(times)-1):
            if levels[i] > levels[i-1] and levels[i] > levels[i+1]:
                highs.append((times[i], levels[i]))
            elif levels[i] < levels[i-1] and levels[i] < levels[i+1]:
                lows.append((times[i], levels[i]))
        return highs, lows

def datetime_to_epoch_hours(dt):
    # hours since 2000-01-01 00:00 UTC
    epoch = datetime(2000, 1, 1, tzinfo=timezone.utc)
    delta = dt - epoch
    return delta.total_seconds() / 3600.0

def format_time(hours):
    h = int(hours) % 24
    m = int((hours - int(hours)) * 60)
    return f"{h:02d}:{m:02d}"

def draw_graph(levels, start_hours, duration, width=50, height=20):
    if not levels:
        return "No data"
    min_lvl = min(levels)
    max_lvl = max(levels)
    range_lvl = max_lvl - min_lvl or 1
    # Sample points
    cols = width
    step = duration / cols
    data = []
    for i in range(cols):
        t = start_hours + i * step
        # find nearest level
        idx = int((t - start_hours) / step)
        if idx >= len(levels):
            idx = len(levels)-1
        data.append(levels[idx])
    min_d = min(data)
    max_d = max(data)
    range_d = max_d - min_d or 1
    grid = [[' ' for _ in range(cols)] for _ in range(height)]
    for i, val in enumerate(data):
        row = int((val - min_d) / range_d * (height-1))
        row = max(0, min(height-1, row))
        grid[row][i] = '*'
    # Add axis labels
    lines = []
    lines.append("Level (m)")
    for r in range(height-1, -1, -1):
        lvl = min_d + (max_d - min_d) * r / (height-1)
        line = f"{lvl:5.2f} |"
        line += ''.join(grid[r])
        lines.append(line)
    # Time axis
    line = "      " + "+" + "-" * (cols-1)
    lines.append(line)
    # Time labels (every 3 hours)
    labels = ""
    for h in range(0, int(duration)+1, 3):
        pos = int(h / duration * cols)
        label = f"{h:02d}:00"
        # Place label
        while len(labels) < pos:
            labels += " "
        labels += label
    lines.append("       " + labels)
    return "\n".join(lines)

def main():
    parser = argparse.ArgumentParser(description="Offline Tide Calendar")
    parser.add_argument("--date", help="Start date YYYY-MM-DD")
    parser.add_argument("--hours", type=float, default=24.0, help="Forecast duration (hours)")
    parser.add_argument("--graph", action="store_true", help="Show ASCII graph")
    parser.add_argument("--list", action="store_true", help="Show high/low tides (default)")
    parser.add_argument("--config", help="JSON file with constituents")
    parser.add_argument("--save-config", help="Save current config to file")
    args = parser.parse_args()

    # Load config
    constituents = DEFAULT_CONSTITUENTS
    mean_level = MEAN_LEVEL
    if args.config:
        with open(args.config, "r") as f:
            data = json.load(f)
            constituents = [tuple(c) for c in data.get("constituents", DEFAULT_CONSTITUENTS)]
            mean_level = data.get("mean_level", MEAN_LEVEL)

    if args.save_config:
        data = {
            "constituents": constituents,
            "mean_level": mean_level
        }
        with open(args.save_config, "w") as f:
            json.dump(data, f, indent=2)
        print(f"Config saved to {args.save_config}")
        return

    # Determine start time
    if args.date:
        dt = datetime.strptime(args.date, "%Y-%m-%d").replace(tzinfo=timezone.utc)
    else:
        dt = datetime.now(timezone.utc).replace(hour=0, minute=0, second=0, microsecond=0)

    start_hours = datetime_to_epoch_hours(dt)
    model = TideModel(constituents, mean_level)
    highs, lows = model.find_extrema(start_hours, args.hours)

    if args.graph:
        # Generate graph
        levels = []
        step = 0.1
        t = start_hours
        while t <= start_hours + args.hours:
            levels.append(model.level(t))
            t += step
        print(draw_graph(levels, start_hours, args.hours))
    else:
        # Table
        print(f"\n🌊 Offline Tide Calendar")
        print(f"Date: {dt.strftime('%Y-%m-%d %H:%M')} – {args.hours}h forecast\n")
        print("High/Low Tides:")
        if not highs and not lows:
            print("No extrema found (check range).")
        else:
            all_events = sorted(highs + lows, key=lambda x: x[0])
            for t, lvl in all_events:
                if t >= start_hours and t <= start_hours + args.hours:
                    dt_event = datetime(2000, 1, 1, tzinfo=timezone.utc) + timedelta(hours=t - datetime_to_epoch_hours(datetime(2000,1,1).replace(tzinfo=timezone.utc)))
                    # Actually we need the date from start
                    date_event = dt + timedelta(hours=t - start_hours)
                    typ = "High" if any(abs(lvl - h[1]) < 0.01 for h in highs) else "Low"
                    print(f"{date_event.strftime('%Y-%m-%d %H:%M')}  {typ:4}  {lvl:5.2f} m")

if __name__ == "__main__":
    main()
